Add the 'edge-ami' image type based on edge-raw-image

7flying opened the pull request:
This PR allows to create an 'edge-ami' image type, which generates an uncompressed raw image (x86_64,
aarch64) ready to upload to AWS EC2 based on an edge OSTree commit.
This edge-ami image type has Ignition support.


- [x] adequate testing for the new functionality or fixed issue
- [ ] adequate documentation informing people about the change such as
  - [ ] submit a PR for the [guides](https://github.com/osbuild/guides) repository if this PR changed any behavior described there: https://www.osbuild.org/guides/

<!--
Thanks for proposing a change to osbuild-composer!

Please don't remove the above check list. These are things that each pull
request must have before it is merged. It helps maintainers to not forget
anything.

If the reason for ticking any of the boxes is ambiguous, please add a short
note explaining why.

In addition, if this pull request fixes a downstream issue, please refer to
test/README.md and add these additional items:

- [ ] 1st commit of any `rhbz#` related PR contains bug reproducer; CI reports FAIL or
- [ ] PR contains automated tests for new functionality and
- [ ] QE has approved reproducer/new tests and
- [ ] Subsequent commits provide bug fixes without modifying the reproducer; CI reports PASS and
- [ ] QE approves this PR; RHBZ status is set to `MODIFIED + Verified=Tested`

Information regarding our GitLab pipeline (Schutzbot):

CI will not be ran automatically if WIP label is applied or the PR is in DRAFT state, instead only
a link will be provided to the pipeline which can then be triggered manually if desired. To run the
CI automatically either switch the PR to ready or apply WIP+test label.

Outside contributors need manual approval from one of the osbuild-composer maintainers.

Schutzbot will only be triggered if all Tests jobs in GitHub workflow succeed.
-->

henrywang commented:
Thanks @7flying! I have some questions about this feature:
1. For igntion, both embemded and firstboot are supported by edge-ami?
2. I didn't find cloud-init package in edge-ami. Does that mean ignition is the only method to add user and setup system?
3. If no cloud-init installed, how to support disk expend.
4. Does this image support both KVM and Xen ec2 instance?

7flying commented:
> Thanks @7flying! I have some questions about this feature:
> 
>     1. For igntion, both embemded and firstboot are supported by edge-ami?
> 
Just firstboot.

>     2. I didn't find cloud-init package in edge-ami. Does that mean ignition is the only method to add user and setup system?
> 

You can add users via ignition or with the blueprint configurations.

>     3. If no cloud-init installed, how to support disk expend.
> 

You can't do it now as part of the installation.

>     4. Does this image support both KVM and Xen ec2 instance?

I'll get back to you on this one. I've used the default settings so I don't know which one is it.

Just a quick note: this PR is basically an uncompressed raw image since we are not going to add cloud-init for the time being. So, we might close this PR and simply ask the user to uncompress the raw image for them to use it on AWS. But @runcom is checking what we should do.

teg commented:
Nice!

Is the ignition support optional? I think it might make sence to make it independent of the image type, in case it is not.

7flying commented:
Fixed conflicts, rebased on main

7flying commented:
Removed the recently deprecated `bootType` field from the image definition options.

7flying commented:
Rebased on main

thozza commented:
How does this image differ from `edge-raw-image`? IOW can't we just import that one into AWS? 🤔

7flying commented:
> How does this image differ from `edge-raw-image`? IOW can't we just import that one into AWS? thinking

This one is simply already uncompressed. 
Yes you can upload the raw image once that you uncompress it. We told that to our PM and the decision was made to have a separate image, that's all.

thozza commented:
> > How does this image differ from `edge-raw-image`? IOW can't we just import that one into AWS? thinking
> 
> This one is simply already uncompressed. Yes you can upload the raw image once that you uncompress it. We told that to our PM and the decision was made to have a separate image, that's all.

composer will decompress the image when importing to AWS, this is transparent to the user and would work. I'm not saying that adding a separate image type is bad or wrong, but I would like us to discuss other options. We can also just add an alias to an existing image type if you really want to have a unique image type to specify on CLI or API.

7flying commented:
> > > How does this image differ from `edge-raw-image`? IOW can't we just import that one into AWS? thinking
> > 
> > 
> > This one is simply already uncompressed. Yes you can upload the raw image once that you uncompress it. We told that to our PM and the decision was made to have a separate image, that's all.
> 
> composer will decompress the image when importing to AWS, this is transparent to the user and would work.
> 
> https://github.com/osbuild/osbuild-composer/blob/be6119ccfeb20bab06725400f1393377d83ead81/cmd/osbuild-worker/jobimpl-osbuild.go#L507-L519
> 
> I'm not saying that adding a separate image type is bad or wrong, but I would like us to discuss other options. We can also just add an alias to an existing image type if you really want to have a unique image type to specify on CLI or API.

So, we would make an`edge-ami` alias to `edge-raw-image` so that when composing an `edge-ami` it is always uploaded to AWS, makes sense. But, if for whatever reason I need to download the `edge-ami` rather than directly uploading it to AWS, can I still do that without specifying another image type?

thozza commented:
> So, we would make an`edge-ami` alias to `edge-raw-image` so that when composing an `edge-ami` it is always uploaded to AWS, makes sense. But, if for whatever reason I need to download the `edge-ami` rather than directly upload it to AWS, could I still do that without specifying another image type in the code?

Things are a bit more complicated.

On premise (Weldr API), we do not restrict which upload options you can pass with which image type. I didn't test it, but it should be already possible to trigger an `edge-raw-image` compose with AWS upload options in a provided `TOML` file and composer would try ti import the image to AWS, which should presumably succeed.

In the service (Cloud API), we tie together image types and upload options so in reality, you can upload any image type to only a single upload target. For edge images it is AFAIK S3. So there, the image name alias would actually help out, as it would make it easy to import `edge-ami` to AWS EC2 and not to S3.

An image name alias makes it possible to trigger a compose on premise using the name. So in e.g. 
```
composer-cli compose start BLUEPRINT TYPE [IMAGE-NAME PROFILE.TOML]
```
you could specify `edge-ami` as the `TYPE` and it would work and internally it would build the `edge-raw-image`. The only downside of the alias is that we use it mostly for backward compatibility when renaming image types and as a result, those aliases are not explicitly listed in `composer-cli compose types` output.

In the worst case, it should be IMO enough to do 
```go
edgeAmiImgType := edgeRawImgType
edgeAmiImgType.environment = &environment.EC2{}
```
in the `distro/rhel9/edge.go` and add the type to the distro constructor. It should not be necessary to add new implementation to `internal/distro/rhel9/images.go` or `internal/image/ostree_ami.go` if there are no differences between the `edge-ami` and `edge-raw-image`.

It is even possible to export uncompressed `edge-raw-image` by overriding its `exports` property, e.g.:
```go
edgeAmiImgType := edgeRawImgType
edgeAmiImgType.environment = &environment.EC2{}
edgeAmiImgType.exports = []string{"image"}
```

7flying commented:
@thozza ready for review again (thanks for your patience!), Github tests look green, waiting on what GitLab says

7flying commented:
> Thanks a lot for your work! partying_face
> 
> The changes look good, but I'm wondering if it wouldn't make the most sense to squash all commits into a single one? Specifically because the correct behavior (IOW iot/edge-raw-images should not change and unit tests should pass for any commit if bisecting) is achieved only with the whole patch set.
> 

Squashed the implementation and left the tests and manifests files in separate commits.

> In addition, I thought that you'll also generate some image test manifests for the new `edge-ami` image thinking

Yep, my bad, updated!

thozza reviewed (approved):
Great work, all looks good 👍 

I would like to suggest if it would not make sense to add also an `iot-ami` image in Fedora distro in the spirit of keeping Fedora as the upstream to RHEL. As a follow-up of course, this PR is good to be merged as is from my PoV 😉
